SU2_CFD/src/drivers/CDiscAdjMultizoneDriver.cpp

```diff
@@ -0,0 +1,1012 @@
+                                                                           config_container[iZone]);
+        solver_container[iZone][INST_0][MESH_0][FLOW_SOL]->Momentum_Forces(geometry_container[iZone][INST_0][MESH_0],
+                                                                           config_container[iZone]);
+        solver_container[iZone][INST_0][MESH_0][FLOW_SOL]->Friction_Forces(geometry_container[iZone][INST_0][MESH_0],
+                                                                           config_container[iZone]);
+
+        if(config_container[iZone]->GetWeakly_Coupled_Heat()) {
+
+          solver_container[iZone][INST_0][MESH_0][HEAT_SOL]->Heat_Fluxes(geometry_container[iZone][INST_0][MESH_0],
+                                                                         solver_container[iZone][INST_0][MESH_0], config_container[iZone]);
+        }
+        solver_container[iZone][INST_0][MESH_0][FLOW_SOL]->Evaluate_ObjFunc(config_container[iZone]);
+        break;
+      case DISC_ADJ_HEAT:
+        solver_container[iZone][INST_0][MESH_0][HEAT_SOL]->Heat_Fluxes(geometry_container[iZone][INST_0][MESH_0],
+                                                                       solver_container[iZone][INST_0][MESH_0], config_container[iZone]);
+        break;
+    }
+
+    direct_output[iZone]->SetHistory_Output(geometry_container[iZone][INST_0][MESH_0],
+                                            solver_container[iZone][INST_0][MESH_0], config_container[iZone]);
+  }
+
+  /*--- Extract objective function values. ---*/
+
+  for (iZone = 0; iZone < nZone; iZone++){
+
+    nMarker_Analyze = config_container[iZone]->GetnMarker_Monitoring();
+
+    for (iMarker_Analyze = 0; iMarker_Analyze < nMarker_Analyze; iMarker_Analyze++) {
+
+      Weight_ObjFunc = config_container[iZone]->GetWeight_ObjFunc(iMarker_Analyze);
+
+      switch (config_container[iZone]->GetKind_Solver()) {
+
+        case DISC_ADJ_EULER:          case DISC_ADJ_NAVIER_STOKES:          case DISC_ADJ_RANS:
+          // per-surface output to be added soon
+          break;
+
+        case HEAT_EQUATION_FVM: case DISC_ADJ_HEAT:
+
+          // per-surface output to be added soon
+          break;
+
+        default:
+          break;
+      }
+    }
+
+    /*--- Not-per-surface objective functions (shall not be included above) ---*/
+
+    Weight_ObjFunc = config_container[iZone]->GetWeight_ObjFunc(0);
+
+    switch (config_container[iZone]->GetKind_Solver()) {
+
+      case DISC_ADJ_EULER: case DISC_ADJ_NAVIER_STOKES: case DISC_ADJ_RANS:
+      case DISC_ADJ_INC_EULER: case DISC_ADJ_INC_NAVIER_STOKES: case DISC_ADJ_INC_RANS:
+
+        switch(config_container[iZone]->GetKind_ObjFunc()) {
+
+          // Aerodynamic coefficients
+
+          case DRAG_COEFFICIENT:
+            ObjFunc+=direct_output[iZone]->GetHistoryFieldValue("DRAG")*Weight_ObjFunc;
```

Comment:
I didn't consider the weighted objective as this will in fact be moved to the new output classes (you are referring to the `Total_ComboObj` variable that is part of the solvers, right?) some day.
This driver-level routine just sums up contributions from different zones.

Reply:
I see... I know it is not an easy problem to solve, I thought about it once and gave up. But since this will be moved to output classes would it make sense to start moving this logic there?